### PR TITLE
Merge matured segmenting transducer to master branch

### PR DESCRIPTION
**Pull request by philip30**
This is the start of merging matured segmenting transducer to the master branch. This should not ready for review yet, but I'll be working to make it work.

### CODE REVIEW

**Comment by philip30** [comment]
This should be ready for review (@neubig , @msperber ). A lot of additions other than segmenting transducer itself, it summed some of my works:
1. Make the loss is not only a float value, but a dictionary. This is to make a clear break down if the loss is composed of different sources.
2. Fixed the attention reporting when we are doing beam search. But my attention matrix is trg in the row, this might be an issue for (@pierregodard )
3. Added global_fertility and attention_entropy loss type in the translator.
4. Some refinement in the LossBuilder. Now XNMT expect LossBuilder object to be the return value of calc_loss(). It supports some basic interface like a standard computational graph, but it is just actually a dictionary of losses. 
5. Model needs to define what are their primary loss (to be picked from LossBuilder) to be used as a primary development loss score. "mle" is the standard for Translator.
6. Added a working segmenting_transducer

**Comment by philip30** [comment]
All test have passed expect the last one https://travis-ci.org/neulab/xnmt/builds/336714950#L2269
I am not sure what is wrong with this, can @msperber help? Thanks!

File: xnmt/search_strategy.py [hunk 1]
```diff
@@ -32,8 +33,9 @@ def generate_output(self, decoder, attender, output_embedder, dec_state, src_len
 
       score += logsoftmax[cur_id]
       word_ids.append(cur_id)
+      attentions.append(attender.get_last_attention())
 
-    return word_ids, score
+    return {"word_ids": word_ids, "attentions": attentions}, score
```
Comment (msperber) [inline xnmt/search_strategy.py]: This is just a suggestion and could also be done later, but Python offers collections.namedtuple for a bit cleaner handling of these situations.

**Comment by msperber** [comment]
Regarding the failing test: it might be best to wait until #274 is merged, maybe that would already solve the problem because it fixes a few bugs with the deserializer.

File: xnmt/__init__.py [hunk 2]
```diff
@@ -4,3 +4,15 @@
 package_dir = os.path.dirname(os.path.abspath(__file__))
 if package_dir not in sys.path:
   sys.path.append(package_dir)
+
+# Initialize the yaml parser by importing everything here
```
[the remark this answers is not in the document]
Comment (philip30) [reply xnmt/__init__.py]: I think the most standard way is to use __init__.py as it is in the usual python packages. But I will follow the convention for now, and maybe will refactor it later.

File: test/config/segmenting.yaml [hunk 3]
```diff
@@ -41,8 +64,10 @@ debug-segmenting: !Experiment
     decoder: !MlpSoftmaxDecoder
       layers: 1
       mlp_hidden_dim: 16
-      bridge: !CopyBridge {}
-    inference: !SimpleInference {}
+      bridge: !CopyBridge {}  
+    inference: !SimpleInference
+      report_path: examples/output/<EXP>.report
```
[the remark this answers is not in the document]
Comment (philip30) [reply test/config/segmenting.yaml]: Done!

File: xnmt/generator.py [hunk 4]
```diff
@@ -20,6 +20,10 @@ def initialize_generator(self, **kwargs):
   def new_epoch(self, training_regimen, num_sents):
     pass
 
+  @register_xnmt_event
+  def next_epoch(self):
+    pass
+
```
[the remark this answers is not in the document]
Comment (philip30) [reply xnmt/generator.py]: This is actually the end of the epoch. I will rename it for clarity.

[the remark this answers is not in the document]
Comment (philip30) [reply xnmt/generator.py]: Oh I think I want to access that epoch counter, so I can remove my own epoch counter. How do I do that?

File: xnmt/inference.py [hunk 5]
```diff
@@ -52,7 +52,7 @@ def __init__(self, model_file=None, src_file=None, trg_file=None, ref_file=None,
     self.len_norm_type = len_norm_type
     self.mode = mode
     self.batcher = batcher
-    
+
```
[the remark this answers is not in the document]
Comment (philip30) [reply xnmt/inference.py]: I actually removed the unnecessary whitespace characters by executing the script ``script/code/clean.sh [source_code]`` with a for loop. So It is not automatic but manual. I can refrain to do this, but it will be better if we do the cleaning every time before pushing the code.

File: xnmt/input.py [hunk 6]
```diff
@@ -26,8 +26,9 @@ class SimpleSentenceInput(Input):
   """
   A simple sent, represented as a list of tokens
   """
```
[the remark this answers is not in the document]
Comment (philip30) [reply xnmt/input.py]: This is actually a very special case for the WordOnlySegmentTransformer where I need to take a look at the external vocab, and so I need to turn back the word ids to its word. Can you explain how to use the Ref(Path())? Or maybe we can just add option to include the original words in the sentence along with the word ids (it will increase the memory use, but will be much simpler).

File: xnmt/search_strategy.py (hunk 1 above)
Comment (philip30) [reply xnmt/search_strategy.py]: Yeah I can do this. This shouldn't be too much problem.

File: xnmt/generator.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (philip30) [reply xnmt/generator.py]: I figured this out.

**Comment by neubig** [comment]
I merged #274 to fix the remaining test, but this unfortunately causes a significant number of merge conflicts with this. Sorry @philip30, but we'll probably have to manually merge these.

**Comment by philip30** [comment]
@neubig No problem, I'll fix that.

**Comment by philip30** [comment]
OK, I have fixed everything.

File: xnmt/translator.py [hunk 7]
```diff
@@ -191,46 +244,29 @@ def html_report(self, context=None):
...
+    with io.open(self.get_report_path() + ".attention.txt", encoding='utf-8', mode='w') as attn_file:
+      for i in range(len(src)+1):
+        if i == 0:
+          words = trg + [""]
+        else:
+          words = ["%.4f" % (f) for f in attn[i-1]] + [src[i-1]]
+        str_format = ""
+        for length in col_length:
+          str_format += "{:%ds}" % (length+2)
+        print(str_format.format(*words), file=attn_file)
```
Comment (pierregodard) [inline xnmt/translator.py]: Hi Philip! Commenting here re: this change breaking things for me. If we merge it this way, I indeed will have to reintegrate somehow on a local branch the code I previously wrote, which is not a big deal. But just two thoughts:

* Shouldn't a report of the attention matrices be more agnostic of the particular formatting that either you or I prefer? Say using `numpy.savetxt` after adding a column and a line to have the source and target labels? If we just want to look at them it might be less readable, but to post-process them, it would be easier to do it this way in order to load the matrices in an external script. What do you think?
* My case for writing matrices where lines are indexed by the target and columns are indexed by the source, and that sum to 1.0 line-wise, is that it seems the standard convention, and certainly the convention in [(Bahdanau, 2014)](https://arxiv.org/pdf/1409.0473.pdf).

Comment (philip30) [reply xnmt/translator.py]: Hi @pierregodard, thank you for the comment. I think the purpose of reporting is to provide users with an easy readable input, not to provide some formatting. The post-processable output should be a type of output, not a report. So adding one more field "attn_file" in addition to "hyp_file" in the translator is more appropriate I think. And you are right, it should be saved in binary format so that there is no loss of digit cancellation because floats cant be fully represented by string.

I think it is easier to sum it down, rather than to the right, but of course I can add option to make it line wise, probably tomorrow I will add this feature.

File: xnmt/input.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (msperber) [reply xnmt/input.py]: If you need the vocab when you're handling sentences inside the AnnotatedSentenceInput, you could just add a vocab field to```AnnotatedSentenceInput.__init__()```, and then in the config file set it as something like 
```
!WordOnlySegmentTransformer:
  vocab: !Ref { path: model.src_reader.vocab }
  ...
```
(or same with trg_reader if it's on the target side). If that solves the problem, I think that would be cleaner because it doesn't require any changes to the SentenceInput code...

File: xnmt/parameters.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
+
```
[the remark this answers is not in the document]
Comment (msperber) [reply xnmt/parameters.py]: Ok, got it! Sounds like a nice idea, maybe I'll start using this myself.. One caveat is that I think when saving the model and reloading it, this will produce unexpected results because it will set the values back to the original ones. If you want to fix that, you could do something like ```self.serialize_params['initial']=new_value``` whenever updating the value. Then, when it will be loaded the next time, the updated value will be passed as initial.

Or, if you think that's better, you can make it a bit cleaner by having both an ```initial``` and a ```current``` value, and using the second one if given. Then, it can be used for serialization by calling ```self.serialize_params['current']=new_value```

File: xnmt/input.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (philip30) [reply xnmt/input.py]: This is a good idea! I wonder why didn't I think about this before. Thanks!

File: xnmt/parameters.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (philip30) [reply xnmt/parameters.py]: this is a good idea. I will do and test this. BTW, does the epoch_counter persist through the separate runs?

File: xnmt/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (msperber) [reply xnmt/__init__.py]: Sure, please feel free to refactor this later!

File: xnmt/inference.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (msperber) [reply xnmt/inference.py]: I see, thanks for explaining!

File: xnmt/parameters.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (msperber) [reply xnmt/parameters.py]: No, the epoch counter does not persist through runs. I think when finetuning after a pretraining phase, it makes more sense to reset the counter because we might have different data sizes etc, and it's better to define a completely new training schedule for that. But I had been thinking of making the training state persistent so that we can resume crashed trainings. That's why I added the training state here: https://github.com/neulab/xnmt/blob/master/xnmt/training_task.py#L323
But this is not implemented yet.

**Comment by philip30** [comment]
I think the other requested features can be an addition to the future pull request. Currently, nothing should breaks anything that is already working. I'd advise to merge this and create a future pull request to accommodate:

1. Hyperparameter persistence
2. Attention report transpose
3. Fixes to the external dictionary for WordOnlySegmentTransformer.

**Comment by neubig** [comment]
@msperber if it's OK with you, we can merge.

**Comment by msperber** [comment]
Sure, that's fine with me!
